### ISSUE TITLE
[CN] Make `rems-project/cn-tutorial` a submodule of `rems-project/cerberus`? 

### ISSUE THREAD

**Issue by septract** [rems-project/cerberus#475]
Suggested by @yav as a temporary measure until we manage to split off a separate CN repo containing both the code and CI examples. 

@yav I didn't fully understand why this would be better than the current state, can you explain?

**Comment by yav** [comment]
The current setup is as follows:
  * Tests reside in the tutorial repo
  * CI always tests tests things against the `master` branch of this repo

This is a problem because it is not possible to make changes to Cerberus+CN and the tests and have CI report the correct current state of the code.  For example, consider the following very common situation:
  1. I make a change to CN to fix something in a PR
  2. This breaks the CI, because some test that was previous failing now works
  3. I make a PR in the tutorial repo to fix the test
  4. The CI on my branch would still be showing as broken, because (a) it has no way to know that the tutorial repo has changed, even if it did it'd be running it against the wrong branch (i.e., `master` instead of PR)

This is a bad state, because when someone comes to review my PR they have no way to know if the tests are working or not, and generally we don't want to waste time reviewing PRs that don't at least pass the tests.  You might think that I could have merged the PR in (3) and manually re-run the CI on my PR.  This would fix the CI on my PR, but break everyone else's, because they don't have the changes on my PR yet.   This gets even more annoying, when a PR is not just one commit---it is really convenient to always know what's the current state of a piece of code.

There are a few ways to fix this:
  1. Move the tests from the tutorial repo to this one
  2. Make the tutorial a submodule of this repo
  3. Add more complexity to the CI

I think we should do (1) or (2)---we don't want CI to be any more complex than it needs to be (it already has potential for being very complex) and we don't want Cerberus+CN developers to have to muck around with it.

I am ambivalent between (1) and (2) but even if we do (1) there will be some examples in the tutorial that would have to stay there and we probably want to make sure they don't break, so I'd say we should also do (2) anyway.

Why does (2) help?  A submodule is just a reference to a particular commit in a foreign repo, so you have a standard why to get a repository at a fixed state.   This helps us because we can make our CI work against the current tutorial submodule, and since each branch can have its own pointer to the version of the tests it needs, it will run CI correctly. So in the above example, I'd have a new step "(4) update the submodule on my branch to point to the PR I made".  This will force CI to re-run and it will run with the correct version of the tests, so everything is always in sync.

The only potential slight downside I can see is that if someone adds additional tests to the tutorial, they won't be automatically picked up by Cerberus+CN, instead someone would have to make a PR to this repo to bump the submodule.
Depending on how you look at it, this is actually a good thing: we probably don't want changes to another repo to break the CI automatically.  With the manual process, when we do the bump we can see if tests are failing and if so either fix things on the PR that bumps the submodule, or make relevant tickets, and wait on bumping the submodule until these are resolved.

EDIT: I forgot to mention that the problems described above would also happen if a PR wanted to add some additional tests, which we should probably be doing when we add new features or fix things.

**Comment by dc-mak** [comment]
I think adding CN Tutorial to the Cerberus repo would add to much noise to the Cerberus repo, and same for submodules, so here's a version of 3 which I don't think is to insanely complicated, and will be quite useful eventually for a more complex CI (which we will need because we want to do executable test generation tests, proof tests, and lemma tests on the same files, and likely have the flexibility to have different command variants for each).

I thought it might be possible to do this based on git versions or dates but a little thought shows both of those can't work in the presence of frequent rebasing, as is encouraged. So let's take the division case as an example:

1. Make `cn --in-progress` output the below, and push to `cerberus/master`. CI remains unbroken.
```json
$ cn --in-progress
{ "division": false }
```

2. Change `cn --in-progress` to show 
```json
$ cn --in-progress
{ "division": true }
```

3. Develop feature as normal. CI may be broken during this.

3. When it comes time for review, note all the CI examples which are breaking. In `cn-tutorial/main`, make a patch for the changes to the repo and push to a top-level `in-progress/division.patch` directory. The CI script should apply the patch if `cn --in-progress` shows `{ division: true }` but otherwise skip it. Hence, CI remains unbroken.

<details>
<summary> example patch </summary>

``` diff
diff --git a/src/example-archive/c-testsuite/broken/error-crash/00064.err125.c b/src/example-archive/c-testsuite/broken/error-crash/00064.err125.c
new file mode 100644
index 0000000..177f8d4
--- /dev/null
+++ b/src/example-archive/c-testsuite/broken/error-crash/00064.err125.c
@@ -0,0 +1,13 @@
+/*
+cn: internal error, uncaught exception:
+    Failure("todo")
+*/
+// Cause: `/` operator
+
+#define X 6 / 2
+
+int
+main()
+{
+       return X - 3;
+}
diff --git a/src/example-archive/c-testsuite/working/00064.working.c b/src/example-archive/c-testsuite/working/00064.working.c
deleted file mode 100644
index 2b3d701..0000000
--- a/src/example-archive/c-testsuite/working/00064.working.c
+++ /dev/null
@@ -1,7 +0,0 @@
-#define X 6 / 2
-
-int
-main()
-{
-       return X - 3;
-}
diff --git a/src/example-archive/dafny-tutorial/working/binary_search.c b/src/example-archive/dafny-tutorial/broken/error-crash/binary_search.c
similarity index 100%
rename from src/example-archive/dafny-tutorial/working/binary_search.c
rename to src/example-archive/dafny-tutorial/broken/error-crash/binary_search.c
diff --git a/src/example-archive/java_program_verification_challenges/working/00004_exceptions.c b/src/example-archive/java_program_verification_challenges/broken/error-crash/00004_exceptions.c
similarity index 100%
rename from src/example-archive/java_program_verification_challenges/working/00004_exceptions.c
rename to src/example-archive/java_program_verification_challenges/broken/error-crash/00004_exceptions.c
diff --git a/src/example-archive/simple-examples/working/division.c b/src/example-archive/simple-examples/broken/error-crash/division_crash_1.c
similarity index 100%
rename from src/example-archive/simple-examples/working/division.c
rename to src/example-archive/simple-examples/broken/error-crash/division_crash_1.c
```

</details>

Note that the CI script should error if there's a patch in the overrides that is not mentioned in the list of features mentioned by `cn --in-progress`, but it's fine (and necessary) to allow for features in `--in-progress` which don't have a patch.

4. Iterate on the PR, including any further changes to the CN tutorial patch, until CI is green.

5. Push the PR on the Cerberus repo. CI remains unbroken.

6. Apply and delete `in-progress/division.patch` in `cn-tutorial/main` and push. CI remains unbroken.

7. Remove `{ division: true }` from `cn --in-progress`

It's a bit ugly in terms of the repo history (minimum three commits on the Cerberus repo, and two on the cn-tutorial repo per feature), so I welcome other suggestions.

But it's easy enough to do in Python with `difflib` ([see this](https://github.com/rems-project/c-tree-carver/blob/main/cpp/test/run_test.py)). And as I mentioned earlier, I think it would set us up well for making the CI able to handle running different commands for the same file (sort of like a `compile_commands.json` file for Clang). `difflib` also makes it easy to implement features like diffing the whole output of a test (not just the return code) and also doing auto-promotions like `dune promote` (again ssee the example mentioned earlier).

**Comment by yav** [comment]
What's noisy about having a sub-module?   Apart from being quite complicated, I think this approach would fail in a few different ways:
  1. I don't think we should be merging with `master` without the CI passing, even for small changes---mistakes happen, I've broken the CI in cases when I thought I just changed comments. 
  2. I don't see what's the point of modify the source code to CN in this case---if we wanted to change the CI, we could just have a file that specifies what branch of the tests to check out.  However, that's *exactly* what a sub-module is, so we'd be just implementing it ourselves for no reason.
  3. I really don't think we want to be managing changes to the tests by manually manipulating diffs---we already have revision control, let's use it!
  4. It's much nicer to have a setup where you don't need to make any changes to the code between the time when the CI runs and the code is merged.

**Comment by dc-mak** [comment]
1. I'm confused: where is there merging in `master` without the CI passing? Avoiding that is an explicit goal of the setup.
2. Having a `cn --in-progress` is just a way to communicate across the two repos, and it allows for different branches/in flight changes for different tests and features.
3. This is fair - I think branches may be more sensible, and I think I'm beginning to see your point. 
4. Also confused: where are there any changes to the code between the time when the CI runs and the code is merged? Each step is atomic, and each merge is a passing CI.

## Clarification of workflow with submodules
Regardless, it would have be much clearer if the workflow steps for different kinds of updates to the Cerberus and the CN tutorial repos, were spelled out, because ultimately that's what we developers will have to do and the means by which we want to achieve it are only instrumental/technical details.

## Ensuring submodule updates don't "go backwards"
In particular, I would like to understand the workflow for the case when new tests are added to the `cn-tutorial/main` repo, `cerberus/master` is updated to point to that commit, but there is a `cerberus/feature-branch` PR with a pointing to a `cn-tutorial/feature-branch` which ends up being behind `cn-tutorial/main` because of aforementioned update. **Essentially, is there a way to check/guarantee automatically that updates to the submodule are only fast-forwards?** If so, I would be happy to proceed.

**Comment by yav** [comment]
On (1) and (4): perhaps I misunderstood the idea, but it seems like you need to change something to indicate what's in progress (i.e., steps (1) and (7)) of the original suggestion.

If we choose to use a sub-module for the tests, the setup would be like that:

  * We'd modify the CI to run the tests from the sub-module, instead of manually checking them out from the tutorial.
  * The workflow for making a PR would be:
     1. Make a branch an do your work as usual.  Every time you push to the branch the CI will run the tests and report the current state.
     2. If you need to change the tests, either to fix them, or add new ones for your feature:
        2.1 Modify the tests in the tutorial repo, either on a branch or on `master`, it doesn't really matter, we can decide this in accordance to what would work best for the tutorial.
        2.2 Update the sub-module on your branch to point to the commit with your changes
     3.  When your PR is ready (i.e., tests pass, reviews are approved, etc.) merge it in `master`.
   * On the concern about "going backwards": whenever two branches update a sub-module, this is always a conflict that has to be resolved manually.   So as long as someone does not explicitly decide to go backwards this shouldn't be a problem.   I can think of cases where you *do* want to go backwards, but if we are really worried about this, we can likely modify the CI to also check that the submodule of `master` is always an ancestor (or equal to) to the submodule of the current branch---it's something `git` should know.
  * If we adopt this approach, CN *will not* be tracking the tutorial's `master` automatically.  Instead, if tests are being added to the tutorial independent of CN, the CN devs would have to manually bump the sub-module every now and then using the workflow outlined above.  It is good to do this manually, because it is possible that these tests break the CI, in which case we should probably make tickets to resolve the issues and delay bumping the sub-module until these are addressed.

**Comment by cp526** [comment]
If you all think the submodule option solves our problems, I'm happy with that plan. Don't think I know enough git or github to foresee whether this leaves any problems, but it would be good to have a smoother CI setup.

Kayvan would also be happy with that route, so long as we arrange it that non-CN Cerberus users don't have to check out all the tests from the submodule.

**Comment by septract** [comment]
Okay, I'll talk to @yav and see what would be involved in doing this for real. 

Just as a note: I don't think the submodule idea removes the need to decouple cerberus and CN into different repos. We might want to merge the non-tutorial test examples into the new `rems-project/cn-verifier` repo (or whatever we call it).  But even then we probably want to keep the tutorial in a different repo, and perhaps include it as a submodule.

**Comment by bcpierce00** [comment]
Why keep the tutorial someplace else?

**Comment by dc-mak** [comment]
Please correct if any of this wrong, this is my understanding so far.

## Current process
1. Start work on a fork.
2. Note tutorial files which need changing.
3. Open PR.
4. Temporarily delete changed files from cn-tutorial.
5. Review and merge.
6. Re-add changed files to cn-tutorial.
7. 
The main downside of this is you can't easily review the changes to the tutorial at the same time (one would have manually review a fork of the tutorial).

## New process
1. Start work on cerberus fork.
2. Point submodule to a fork of cn-tutorial (and make changes there).
3. Open PR.
4. Review both.
5. First merge cn-tutorial fork into cn-tutorial.
6. Update submodule to point to cn-tutorial (not the fork) again.
8. Final checks and merge on cerberus side.

We could skip a couple of steps changing cn-tutorial directly instead of via forks. I think this is (a) bad practice and (b) vulnerable to conflicts with two in-flight features. But if we work with forks, I don't know how bisect will hold up.

**Comment by bcpierce00** [comment]
How long is it going to be before we can actually split out CN from
Cerberus?  If not too long, then perhaps the energy that's going into this
discussion (and that's going to go into implementing and teaching people
about any of the proposed solutions) is not optimally utilized?

**Comment by septract** [comment]
Yes, good point @bcpierce00 . A lot of the complexity here would just go away if we just put all CN artifacts into the same repo (which hopefully will happen soon?). 

The submodule idea seemed to me like it might be uncontroversial as a short-term fix. But based on the thread above, obviously there are significant disagreements about it. I'm not sure it's worth the effort of achieving convergence  if in fact a repo split is imminent.

**Comment by septract** [comment]
So I propose:  
* Someone in the know gives us an update on the timeline to a repo split (@dc-mak / @cp526 ?) 
* If it's not too far off, we close this ticket as "won't fix"

**Comment by dc-mak** [comment]
@kmemarian is working on it, but I think we should assume it will take a while/not wait on it.

**Comment by dc-mak** [comment]
@kmemarian can I/someone proceed with this?
